### ISSUE TITLE
AutoNumeric ignores null and undefined values in 'set' method.

### ISSUE THREAD

**Issue by Gle2nn** [autoNumeric/autoNumeric#446]
### Expected behavior
I would expect that the 'current value' would be set to the method parameter value.

### Actual behavior
The parameter value is ignored, and the current value is not changed.

### Steps to reproduce the problem
Here's a very very simple fiddle for your enjoyment: https://jsfiddle.net/Gle2nn/dkw7y66s/

Thanks for reporting a bug or giving us great ideas with your feature request!
You're welcome!

**Comment by BobKnothe** [comment]
@Gle2nn,

Thanks for the fiddle.

Yes thy null and undefined are ignored. So in you opinion what should the 'set' method return in those cases?

**Comment by AlexandreBonneau** [comment]
Thanks for taking the time to create an account and report this @Gle2nn.

Ignoring `null` and `undefined` is currently the wanted behavior;

- `undefined` is obvious since it usually means the dev did something wrong and forgot to initialize his variables, and
- `null` is ignored as well since many functions returns `null` as their default value when they fail.

That's why if you want to 'clear' your autoNumeric element, with the current *(but old)* stable version you need to `set` the empty string `''`.

With the upcoming `v4`  found in the `next` branch, you can just use the more semantic:

```js
anElement.clear();
```

Regards

**Comment by Gle2nn** [comment]
Hi, Bob.

Thanks for responding. You may recall that I reported a similar problem 
with null in 1.9.45.

To answer your specific question, to be honest I'm surprised that a 
mutator such as 'set' would return any value at all. This is the only 
documentation I could find on the set method:

|set| 	Set the value given as a parameter, and formats it 
|$(someSelector).autoNumeric('set', '12345.67');|


If an error occurs, I think an appropriate action would be to throw an 
exception.

It may help if you have some visibility on my use case. I am trying to 
provide a simple UI for a nullable SQL database column. Therefore, I 
need to support user entry of null values (create/edit), and I need to 
be able to pass null values to the UI (edit).

A more concrete example might be a database table containing weather 
data for a particular location and time, such as temperature, wind 
speed, precipitation, etc. If it doesn't rain or snow on a particular 
day, I would want to record the total precipitation for that day as 
null, since there was no precip. I could store a zero, but that makes it 
harder to determine the average precipitation amount for days that 
actually had rain/snow.
Another case would be temperature. Let's say the automated recorder 
fails on a particular day. In this case, the temp would be recorded as 
null, since no reading was taken (zero wouldn't be appropriate for this 
case).

So the basic behaviours I would like to see as an end-user:

  * A field could be set to undefined (no user entry has been made for
    this field) or null (a null value was retrieved from the db).
  * Tabbing through a field without making any changes shouldn't change
    a null or undefined value.
  * Removing all the text from a field should set the value to null.

Thanks again
Gle2nn

**Comment by Gle2nn** [comment]
Alexandre:

My previous response describes a use case that explains why null and 
undefined shouldn't be ignored.

undefined isn't always an indicator of a dev mistake. For example, the 
JSON serializer I use on the server allows me to suppress fields with a 
null value.
On the client, a statement such as
     _temp = ko.observable( pojo.temperature );
is used to store the value in a KnockoutJS observable. If the 
temperature is null on the server, then pojo.temperature is undefined on 
the client.
Most code will treat undefined and null the same way (ie. _temp() == 
null is true if the value is undefined or null).

With regards to null, I can understand that you are running up against 
your own coding practices and API definitions. As I dev, I would prefer 
to see exceptions in the case of error.
I spent most of a day trying to debug my own code trying to trace this 
problem down (I assumed that I was making the mistakes).
The documentation that I could find does not say anything about ignoring 
null or undefined, or how errors are communicated to the caller.
At least if an exception was thrown I'd have some clue to use to resolve 
the problem.

Cheers,'
Gle2nn

**Comment by AlexandreBonneau** [comment]
Hey @Gle2nn, I'm sorry to read that you had to debug for long time just to find that the culprit was just a single `if` ; I feel you :)

I'll try to answer the points of both your last messages.


`set()` does not return any relevant value ; In the `next` branch, functions can be chained which means `set` (and most other functions) *do* return `this` in order to be able to do something like that for instance:

```js
anElement.french()
         .set(42)
         .update({ options })
         .formSubmitJsonNumericString(callback)
         .clear();
```

About your feature requests :


> Removing all the text from a field should set the value to null.

Since JS is pretty limited when it comes to number manipulation (hello `0.1 + 0.2`), we always keep the autoNumeric value as a string, as much as possible.
We also have options to keep the currency sign when the input is empty, and so far since we only manipulate strings, we decided to keep an empty string when the input value was deleted (and no currency nor suffix text were kept).

That being said, I could see a modification of the `emptyInputBehavior` option to keep `null` instead of `''`. @BobKnothe, what are your thoughts on that?


> Tabbing through a field without making any changes shouldn't change a null or undefined value.

If the new option value `emptyInputBehavior` is set to `null`, then indeed tabbing into that field should not change its `rawValue`.


> A field could be set to undefined (no user entry has been made for this field) or null (a null value was retrieved from the db).

I still think allowing to set `undefined` is just the opening the door to numerous bugs for our users.


So, to sum up ;
- I could see modifying `emptyInputBehavior` so that `null` could be set when emptying the autoNumeric element
- Instead of silently dismissing `undefined` and `null` when used with `set()`, we could show a warning in the console for `null` (if `emptyInputBehavior` is not set to `null`), and throw an error for `undefined`.

Would that work for you?

If yes, please create a new feature requests for modifying the `emptyInputBehavior` so that is sets the value to `null` on empty!

**Comment by Gle2nn** [comment]
Hi, Alexandre:

Thanks for the clarifications and additional discussion points.

Allowing chaining makes sense in certain circumstances such as the one 
you've described. I'd use it in my knockout binding, for example.

I think your proposal for null handling would allow me to do most of 
what I want to do with AutoNumeric, and I will create a change request 
based on this thread. One change I would propose would be to handle null 
or undefined as an empty string if the new option is not in play.

If you plan to support null as described, I suspect supporting undefined 
will come along for (almost) free due to coercion.

I was hoping to use undefined as a value to allow me to control whether 
a defaulting rule could be applied to the field. I think using undefined 
in this case makes sense semantically (i.e. the user has not entered any 
value). Remember that null is a legitimate value in my use case, so I 
cannot use null. I will proceed with using a value of a different type.

I understand and agree with your perspective on the use of undefined. 
Maybe a compromise would be to add an option that allows the use of 
undefined in AN.

Getting back to the use case, undefined is just a marker that allows me 
to track whether the USER changed the field as opposed to the form. Is 
there a way to get that information from AN?

Thanks for your time,
Gle2nn

**Comment by AlexandreBonneau** [comment]
> Getting back to the use case, undefined is just a marker that allows me to track whether the USER changed the field as opposed to the form. Is there a way to get that information from AN?

Yes, using the native `change` event
